# Project 1
## Rachel Jordan Input 

In [1]:
library("igraph")


Attaching package: ‘igraph’

The following objects are masked from ‘package:stats’:

    decompose, spectrum

The following object is masked from ‘package:base’:

    union



Load the data 

In [2]:
load("/Users/rachelcjordan/Downloads/astrocollab.Rdata")

In [3]:
summary(astrocollab)

IGRAPH UNW- 16706 121251 -- Astrophysics collaborations
+ attr: name (g/c), Author (g/c), Citation (g/c), URL (g/c),
| Description (g/c), name (v/c), weight (e/n)


This is a undirected, named, and weighted graph this has 16706 vertices and 121251 edges attributes of the graph include: name, author, citation, URL, description, name, weight


Data Prep;  #captures the vertices in order 
#16706/16706 vertices

In [4]:
V(astrocollab)

This graph was created by an old(er) igraph version.
  Call upgrade_graph() on it to use with the current igraph version
  For now we convert it on the fly...
This graph was created by an old(er) igraph version.
  Call upgrade_graph() on it to use with the current igraph version
  For now we convert it on the fly...


+ 16706/16706 vertices, named:
    [1] BIERMANN, PL                     STANEV, TKGT                    
    [3] GOLDMAN, I                       WANDEL, A                       
    [5] PILDIS, RA                       BREGMAN, JN                     
    [7] EVRARD, AE                       SWARTZ, DA                      
    [9] SUTHERLAND, PG                   HARKNESS, RP                    
   [11] BECKER, W                        BARZIER, KTS                    
   [13] TRUEMPER, J                      ANGELANTONJ, C                  
   [15] AMENDOLA, L                      LITTERIO, M                     
   [17] OCCHIONERO, F                    LOEB, A                         
   [19] BRAINERD, JJ                     MEEGAN, CA                      
+ ... omitted several vertices

In [5]:
upgrade_graph(astrocollab)

IGRAPH UNW- 16706 121251 -- Astrophysics collaborations
+ attr: name (g/c), Author (g/c), Citation (g/c), URL (g/c),
| Description (g/c), name (v/c), weight (e/n)
+ edges (vertex names):
 [1] BIERMANN, PL  --STANEV, TKGT   GOLDMAN, I    --WANDEL, A     
 [3] PILDIS, RA    --BREGMAN, JN    PILDIS, RA    --EVRARD, AE    
 [5] BREGMAN, JN   --EVRARD, AE     SWARTZ, DA    --SUTHERLAND, PG
 [7] SWARTZ, DA    --HARKNESS, RP   SUTHERLAND, PG--HARKNESS, RP  
 [9] BECKER, W     --BARZIER, KTS   BECKER, W     --TRUEMPER, J   
[11] BARZIER, KTS  --TRUEMPER, J    ANGELANTONJ, C--AMENDOLA, L   
[13] ANGELANTONJ, C--LITTERIO, M    AMENDOLA, L   --LITTERIO, M   
+ ... omitted several edges

In [6]:
myastro = upgrade_graph(astrocollab)

In [7]:
myastro

IGRAPH UNW- 16706 121251 -- Astrophysics collaborations
+ attr: name (g/c), Author (g/c), Citation (g/c), URL (g/c),
| Description (g/c), name (v/c), weight (e/n)
+ edges (vertex names):
 [1] BIERMANN, PL  --STANEV, TKGT   GOLDMAN, I    --WANDEL, A     
 [3] PILDIS, RA    --BREGMAN, JN    PILDIS, RA    --EVRARD, AE    
 [5] BREGMAN, JN   --EVRARD, AE     SWARTZ, DA    --SUTHERLAND, PG
 [7] SWARTZ, DA    --HARKNESS, RP   SUTHERLAND, PG--HARKNESS, RP  
 [9] BECKER, W     --BARZIER, KTS   BECKER, W     --TRUEMPER, J   
[11] BARZIER, KTS  --TRUEMPER, J    ANGELANTONJ, C--AMENDOLA, L   
[13] ANGELANTONJ, C--LITTERIO, M    AMENDOLA, L   --LITTERIO, M   
+ ... omitted several edges

In [ ]:
plot(myastro)

As we can see from above, we have a ton of data.  It not only takes a long time to process in R (if we can process all of the data due to memory capacity, but also the results (example plot above) is incomprehensible. 

Prior to exploring the dataset and using different functions,we need to sample the dataset to get a smaller data size that we can explore 



In [ ]:
is.simple(myastro) 
#result is true (simple) because its a graph without loops and multiple edges between vertices

In [ ]:
myastro <- simplify(myastro) # if needed, but not needed

In [ ]:
plot(myastro) #this is way too closely overlapping and not helpful, so need to take a sample 


In [ ]:
#in order to simplify the amount of data (vertices and edges) we are looking at, lets find degrees

degree <- degree(myastro)
hist(degree)
astrodist = degree.distribution(myastro, FALSE)
astrodist = astrodist*100.0
astrodist[1:50]

#our new dataset is now astronew, as it is a sample of the original myastro dataset 


In [ ]:
#delete all but the first 50 vertices 
astroseq <- V(myastro)
astronew <- myastro - astroseq[51:16706]
astronew <- simplify(astronew)
plot(astronew)

# 3. Experiement with some function that I have shown in the assoicate PPT file on BB. Preesent the results in writeup. 

In [ ]:
#trying to build a standard representation of the graph (subgraph based on simplifying)
vertex_attr(astronew) #finding vertex attriutes 

Next using a function to create and adjacency matrix, which is a square matrix.  

This function will query and find the adjacent vertices for multiple vertices at once 

Finding the adjacency matrix for the original dataset /matrix 


In [ ]:
#adjmatrix = as_adjacency_matrix(myastro) #adjacency matrix


In [ ]:
#adjmatrix #this is too big 

Trying with small subset 

In [ ]:
adjmatrix2 = as_adjacency_matrix(astronew) #adjacency matrix


In [ ]:
adjmatrix2

Another function is the function that adds weights to the edges and vertices 
so in order to subset the matrix or create a subgraph, we can use the weights 

In [ ]:
E(astronew)$weight = rnorm(ecount(astronew))

In [ ]:
V(astronew)$weight = rnorm(vcount(astronew))

In [ ]:
#merge the vertices with positive weight into a subgraph including the edge weight

In [ ]:
sg = induced.subgraph(astronew, which(V(astronew)$weight > 0.4))

In [ ]:
plot(sg, edge.label=round(E(sg)$weight, 5))

Initially received an error, but when used the subset graph, was able to use and complete this function 

Next, I want to try the using the walktrap function which tries to find the connected
subgraphs, called communities, in a graph via random walks 

In [ ]:
wc = walktrap.community(myastro)

In [ ]:
wc

In [ ]:
plot(wc,myastro, vertex.size = .5, layout=layout.fruchterman.reingold)
#groups: 4879, mod: 0.21

In [ ]:
#try to do this with subgraph, astronew
wc2 = walktrap.community(astronew)

In [ ]:
wc2

In [ ]:
plot(wc2,astronew, vertex.size = .5, layout=layout.fruchterman.reingold)
#here we found 23 groups, mod: 0.86

In [ ]:
#another try but specifying number of steps of random walks to perform w/ weights
wc3 = walktrap.community(astronew, weights = E(astronew)$weight, steps = 5)

In [ ]:
wc3

In [ ]:
plot(wc3,astronew, vertex.size = .5, layout=layout.fruchterman.reingold)
#here we found 23 groups, mod: 0.86

Lastly, want to look at the degrees of my graph and subgraphs

In [ ]:
degree(astronew) #here each node in myastro is given number of degrees 
#which is the sum of its in and out degrees. 

In [ ]:
hist(degree(astronew))

# 4. Explore other functions 

#Function 1: 
#as_adj_list:  This created adjacency lists from a graph, either for adjacent edges or for neighboring vertices 

In [ ]:
as_adj_list(astronew, mode = c("all", "out", "in", "total"))

In [ ]:
as_adj_edge_list(astronew, mode = c("all", "out", "in", "total"))

Function 2: gorder()
This function provides the number of vertices of a graph

In [ ]:
gorder(myastro) #16706

In [ ]:
gorder(astronew) #50

Function 3: cluster_infomap

In [ ]:
cluster_infomap(myastro) #this takes way too long for larger dataset

In [ ]:
cluster_infomap(astronew)

Function 4: decompose()

This function creates a separate graph for each component of a graph 

In [ ]:
dg <- decompose.graph(astronew)

In [ ]:
plot(dg[[1]])

Function 5: Betweenness which helps with community detection

In [ ]:
betweenness(myastro)

In [ ]:
eb <- edge.betweenness(myastro)

In [ ]:
plot(eb)

Function 6: Clusters; connected components 

In [ ]:
c1 <- clusters(astronew)

In [ ]:
c1

Function 7: Find neighborhood of vertices using graph.neighborhoods

In [ ]:
gn <- graph.neighborhood(astronew, order =1)

In [ ]:
plot(gn[[1]])

In [ ]:
plot(gn[[2]])

Function 8: eccentricity of the verticies in a graph

In [ ]:
eccentricity(astronew, vids = V(astronew), mode = c("all", "out", "in", "total"))

The function above calculates the exxentricity of a vertex is its shortest path distance from the fathest other node in the graph

Function 9: The girth function measures the length of the shortest circle in a graph

In [ ]:
girth(astronew, circle = TRUE)

Function 10: is_directed()

In [ ]:
is_directed(astronew)

This is a function that can be easily used to determine if a graph is directed or not. 

# Part 5 

In [ ]:
#(a)Centrality

#ac1 = alpha_centrality(myastro)
#ac1



In [ ]:
ac2 = alpha_centrality(astronew)


In [ ]:
ac2
max(ac2) # this is GELLER, DGFMJ, with 4.91417 centrality score 
#this function calculates the alpha centrality of the vertices in the given graph

In [ ]:
#(b)longest path 
#using diameter 
#this function calculates the diameter of a graph which is the length of the longest geodesic (shortest possible path)

dia = diameter(myastro, directed = FALSE)


In [ ]:
max(dia)


In [ ]:
get.diameter(astronew)

#(c) largest clique 

#this function finds all the largest or all of the max cliques in an undirected graph.  
#the size of the largest clique can also be calculated using max_cliques


In [ ]:

max_cliques(myastro)

the result: [[21]]+ 5/50 vertices, named: [1] MEEGAN, CA    BRAINERD, JJ  BROCK, MN     PENDLETON, GN BRIGGS, MS   



(d) ego
This is the functions that find the vertices not farther than a given limit from another fixed vertex, these are called the neighborhood of the vertex finding neighbors

In [ ]:
ego(astronew, 1, nodes = V(astronew), mindist = 1)
ego(astronew, 1, nodes = V(astronew), mindist = 0)



In [ ]:

#(e) power centrality 
#take a graph and returns the Boncich power of centralities of positions
#(selected by nodes) and the decay rate for power contributions is specified 
#by exponent (1 by default)

#high magnitudes indicate slower decays 

power_centrality(astronew, nodes = V(astronew), loops = FALSE, exponent = 0, rescale = TRUE, tol = 0, sparse = TRUE)

#BRIGGS, MS   PENDLETON, GN       BROCK, MN        MOHR, JJ 
#0.036363636     0.036363636     0.036363636     0.018181818



